File: frontend/src/lib/api/types/nutrition.ts
// ─── Nutrition / Ride Fueling ────────────────────────────────────────────────

export interface FuelScheduleEntry {
  time_min: number;
  carbs_g: number;
  hydration_ml: number;
  sodium_mg: number;
  suggestion?: string | null;
}

export interface RideFuelPlan {
  id: string;
  user_id: string;
  activity_id: string | null;
  planned_duration_min: number | null;
  planned_if: number | null;
  pre_ride_carbs_g: number;
  during_carbs_per_hour_g: number;
  during_hydration_ml_per_hour: number;
  during_sodium_mg_per_hour: number;
  post_ride_carbs_g: number;
  post_ride_protein_g: number;
  schedule: FuelScheduleEntry[];
  actual_pre_ride_notes: string | null;
  actual_during_notes: string | null;
  actual_post_ride_notes: string | null;
  actual_water_ml: number | null;
  actual_carbs_g: number | null;
  actual_electrolytes_mg: number | null;
  source: string;
  created_at: string;
  updated_at: string;
}

export interface CreateFuelPlanPayload {
  activity_id?: string;
  planned_duration_min?: number;
  planned_if?: number;
}

export interface FuelActualsUpdatePayload {
  actual_water_ml?: number | null;
  actual_carbs_g?: number | null;
  actual_electrolytes_mg?: number | null;
}

File: backend/app/schemas/nutrition.py
"""Schemas for ride fuel plans."""

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class FuelPlanCreate(BaseModel):
    """Generate a fuel plan for a planned or completed ride.

    Provide activity_id for a synced ride, OR planned_duration_min for a
    future ride. If both are given the explicit values win.
    """

    activity_id: uuid.UUID | None = None
    planned_duration_min: int | None = Field(default=None, gt=0, le=1440)
    planned_if: float | None = Field(default=None, ge=0.4, le=1.3)


class FuelScheduleEntry(BaseModel):
    time_min: int
    carbs_g: float
    hydration_ml: float
    sodium_mg: float
    suggestion: str


class RideFuelPlanRead(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    activity_id: uuid.UUID | None
    planned_duration_min: int | None
    planned_if: float | None
    pre_ride_carbs_g: float | None
    during_carbs_per_hour_g: float | None
    during_hydration_ml_per_hour: float | None
    during_sodium_mg_per_hour: float | None
    post_ride_carbs_g: float | None
    post_ride_protein_g: float | None
    schedule: list[FuelScheduleEntry] | None
    actual_pre_ride_notes: str | None
    actual_during_notes: str | None
    actual_post_ride_notes: str | None
    actual_water_ml: float | None
    actual_carbs_g: float | None
    actual_electrolytes_mg: float | None
    source: Literal["auto", "manual"]
    created_at: datetime
    updated_at: datetime


class FuelPlanActualsUpdate(BaseModel):
    """Log what was actually consumed (post-ride)."""

    actual_water_ml: float | None = Field(default=None, ge=0)
    actual_carbs_g: float | None = Field(default=None, ge=0)
    actual_electrolytes_mg: float | None = Field(default=None, ge=0)
